File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterModelField.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_datastore.types.model

import com.amplifyframework.core.model.ModelField

data class FlutterModelField(val map: Map<String, Any>) {
    // Name of the field is the name of the instance variable
    // of the Model class.
    private val name: String = map["name"] as String

    // Type of the field is the data type of the instance variables
    // of the Model class.
    private val type: FlutterFieldType = FlutterFieldType(map["type"] as Map<String, Any>)

    // If the field is a required or an optional field
    private val isRequired: Boolean = map["isRequired"] as Boolean

    // If the field is an array targetType. False if it is a primitive
    // targetType and True if it is an array targetType.
    private val isArray: Boolean = map["isArray"] as Boolean

    // True if the field is an enumeration type.
    private val isEnum: Boolean = type.isEnum()

    // True if the field is an instance of model.
    private val isModel: Boolean = type.isModel()

    // True if the field is an instance of CustomType (Embedded)
    private val isCustomType: Boolean = type.isCustomType()

    private val isReadOnly: Boolean = map["isReadOnly"] as Boolean

    // An array of rules for owner based authorization
    private val authRules: List<FlutterAuthRule>? =
        (map["authRules"] as List<Map<String, Any>>?)?.map { FlutterAuthRule(it) }

    // Association (if any) of the model
    private val flutterModelAssociation: FlutterModelAssociation? =
        if (map["association"] == null) null
        else FlutterModelAssociation(map["association"] as Map<String, Any>)

    fun getModelAssociation(): FlutterModelAssociation? {
        return flutterModelAssociation
    }

    fun convertToNativeModelField(): ModelField {
        var builder: ModelField.ModelFieldBuilder = ModelField.builder()
            .name(name)
            .javaClassForValue(type.getJavaClass())
            .targetType(type.getTargetType())
            .isRequired(isRequired)
            .isArray(isArray)
            .isEnum(isEnum)
            .isModel(isModel)
            .isReadOnly(isReadOnly)
            .isCustomType(isCustomType)

        if (!authRules.isNullOrEmpty()) {
            builder = builder.authRules(authRules.map { it.convertToNativeAuthRule() })
        }
        return builder.build()
    }
}

File: packages/amplify_datastore/ios/Classes/types/model/FlutterModelField.swift
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Flutter
import Foundation
import Amplify

public struct FlutterModelField {
    
    public let name: String
    public let type: FlutterModelFieldType
    public let isRequired: Bool
    public let isArray: Bool
    public let isReadOnly: Bool
    public let association: FlutterModelAssociation?
    public let authRules: [FlutterAuthRule]?
    
    init(serializedData: [String: Any]) throws {

        guard let name = serializedData["name"] as? String
        else {
            throw ModelSchemaError.parse(
                className: "FlutterModelField",
                fieldName: "name",
                desiredType: "String")
        }
        self.name = name

        guard let inputTypeMap = serializedData["type"] as? [String: Any] else {
            throw ModelSchemaError.parse(
                className: "FlutterModelField",
                fieldName: "type",
                desiredType: "[String: Any]")
        }
        self.type = try FlutterModelFieldType(serializedData: inputTypeMap)

        guard let isRequired = serializedData["isRequired"] as? Bool
        else {
            throw ModelSchemaError.parse(
                className: "FlutterModelField",
                fieldName: "isRequired",
                desiredType: "Bool")
        }
        self.isRequired = isRequired

        guard let isArray = serializedData["isArray"] as? Bool
        else {
            throw ModelSchemaError.parse(
                className: "FlutterModelField",
                fieldName: "isArray",
                desiredType: "Bool")
        }
        self.isArray = isArray
        
        guard let isReadOnly = serializedData["isReadOnly"] as? Bool
        else {
            throw ModelSchemaError.parse(
                className: "FlutterModelField",
                fieldName: "isReadOnly",
                desiredType: "Bool")
        }
        self.isReadOnly = isReadOnly

        if let inputAssociationMap = serializedData["association"] as? [String : Any]{
            self.association = try FlutterModelAssociation(serializedData: inputAssociationMap)
        }
        else{
            self.association = nil
        }

        if let inputAuthRulesMap = serializedData["authRules"] as? [[String:String]]{
            self.authRules = try inputAuthRulesMap.map{
                try FlutterAuthRule(serializedData: $0)
            }
        }
        else{
            self.authRules = nil
        }
    }

    static func deserializeFieldType(type: String) -> Any.Type {
        if type == "String" {
            return String.self
        }
        if type == "Integer" {
            return Int64.self
        }
        if type == "Double" {
            return Double.self
        }
        if type == "Boolean" {
            return Bool.self
        }
        if type == "DateTime" {
            return String.self
        }
        if type == "Time" {
            return String.self
        }
        if type == "Date" {
            return String.self
        }
        // TODO: Serialize all fields when we do configure()
        //        if let enumType = type as? EnumPersistable.Type {
        //            return .enum(type: enumType)
        //        }
        //        if let modelType = type as? Model.Type {
        //            return .model(name: modelType.modelName)
        //        }
        //        if let embeddedType = type as? Codable.Type {
        //            return .embedded(type: embeddedType, schema: nil)
        //        }
        preconditionFailure("Could not create a ModelFieldType from \(String(describing: type)) MetaType")
    }

    private func convertFlutterAuthRules(flutterAuthRules : [FlutterAuthRule]?) -> [AuthRule]?{

        if let flutterAuthRules = flutterAuthRules{
            var result: [AuthRule] = [AuthRule]();

            for flutterAuthRule in flutterAuthRules {
                result.append( flutterAuthRule.convertToNativeAuthRule() );
            }

            return result;
        }
        else{
            return nil;
        }
    }

    public func convertToNativeModelField(customTypeSchemasRegistry: FlutterSchemaRegistry) throws -> ModelField {
        return ModelField.init(
            name: name,
            type: try type.convertToNativeModelField(customTypeSchemaRegistry: customTypeSchemasRegistry),
            isRequired: isRequired,
            isReadOnly: isReadOnly,
            isArray: isArray,
            association: association?.convertToNativeModelAssociation(),
            authRules: authRules?.map{
                $0.convertToNativeAuthRule()
                } ?? [AuthRule]()
        )
    }

}

typealias FlutterModelFields = [String: FlutterModelField]
